### ISSUE TITLE
Provider Review

### ISSUE THREAD

**Issue by bill-rich** [nutanix/terraform-provider-nutanix#175]
Hello! 

My name is Bill, I'm a member of the Terraform team @ HashiCorp.

I’ve taken a look at the provider here and would like to say great work so far! I do have feedback outlined below that I’d like to see addressed before we move on to the next steps. I’m opening this issues as a sort of checklist for tracking remaining items and discussion. The review was done on the git commit 13167bf1042627382a9040d4dfab7c9d24f8b65a.

- [ ] Add descriptions in schema.
- [ ] Not a big deal, but making err variable naming more consistent could help making things more readable. In some places just “err” is used, but in others longer and more specific names are used.
- [x] Handle errors when setting values (Resource_nutanix_category_key.go:78-80 & others)
- [x] Please review cases where TypeList is used to determine if TypeList or TypeSet is more appropriate. TypeList is used when ordering does matter, and TypeSet is used when order does not. If you use a TypeList where the order may change, you can end up with unnecessary diffs. For example, possibly resource_nutanix_virtual_machine.categories?
- [x] There are singular and plural versions of each of the data sources. Are both needed? 
- [x] Resource_nutanix_category_key.go:43 validaste -> validate (In a few other resources also)
- [x] Keep comment format consistent. Eg “// Comment” vs “//comment”
- [x] image.html.markdown:37 Grammar issue “... of a the source image...”
- [x] uuid -> UUID in docs
- [x] Move the schema out of getSchema function and into the resource function. It makes the resource function shorter, but keeping the schema in the resource function keeps your resources more consistent with other providers.
- [x] Is “api_version” required? Are there situations where users would need to be using different API versions for different resources? If not, I think it would be better to make the API version something that is set globally in the provider.
- [x] Make pointer util funcions more descriptive (eg StringPtr instead of String). The reverse functions are more descriptive.
- [x] Resource_nutanix_category_key.go:43 Log message says “name”, but uses Id
- [x] In general, break out smaller functions for in very complex ones. Breaking the functions into smaller peices can make the code easier to read. An good example is at Resource_nutanix_network_security_rule.go: resourceNutanixNetworkSecurityRuleCreate & getNetworkSecurityRuleResources. 
- [x] Resources should have SetId(“”) in Read function when the resource isn’t found. That way if a resource is removed outside of Terraform, it will be be accurately reflected.
- [x] Include additional tests for resources such as updating, importing and checking that correct values are applied.
- [x] It may be better `power_state` from the virtual machine resource as it could cause the configuration to not match the actual state. For example, a host with `power_state` set to off and guest customizations set, may not actually have those customizations applied. Another example is that it affects the ability of provisioners to run. These result in moving away from the philosophy of immutable infrastructure.
- [x] Fill in descriptions for attributes in the documentation. For example, clusters.markdown.
- [x] There is only documentation for `clusters`, but not `cluster` data source.

**Comment by JonKohler** [comment]
Taking care of some of the very low hanging fruit stuff (documentation, formatting tweaks, etc). For heavier weight things, we'll break them out

**Comment by JonKohler** [comment]
Hey @bill-rich - making good progress here, quick question though.

according to the team, the below two items in the review has not been requested from previous providers before. To complete these will be quite time consuming, just wondering if perhaps provider standards have changed in this regard recently? or if our understand of these is incorrect?

-Add descriptions in schema.
-Check error while setting attributes only for structures and arrays

**Comment by bill-rich** [comment]
The descriptions can actually be skipped. It is a nice to have and is used for things like document generation, not necessary.

The error handling on Set hasn't always been enforced in the past since errors coming from Set are pretty uncommon, but in an attempt to stick with with Go best practices, we're trying to handle errors wherever possible.

**Comment by JonKohler** [comment]
ok @bill-rich. We've polished off every single other task besides these two. 

In talking with @thetonymaster, he's working directly with Hashi on other providers, like AzureStack, and he's mentioning that this didn't get enforced there.

Putting description comment aside for a second, for the go errors, those may be challenging to unit test and make sure they actually work, meaning a lot of work for diminishing returns.

Not trying to skip work, just trying to make sure we know what our next steps are.

Would it be possible pick up the review again from this point? or are we blocked on these last two points?

**Comment by thetonymaster** [comment]
@bill-rich 

> The descriptions can actually be skipped. It is a nice to have and is used for things like document generation, not necessary. 

This definitely can be an improvement in the near future

> The error handling on Set hasn't always been enforced in the past since errors coming from Set are pretty uncommon, but in an attempt to stick with Go best practices, we're trying to handle errors wherever possible.

I can understand where this goes, however, my only concern here is the code coverage, and how to improve it with those chances since I don't see an (and succinct) easy way to test each of the variations, and most of the operations are `Set`.  Granted, this happened before with non-scalar values, however since now all of them have checks it can probably have an impact on coverage.

**Comment by bill-rich** [comment]
@JonKohler @thetonymaster I see the concern there with sacrificing code coverage for error checking on scalar values, and appreciate you working through the balance with me, but I think it is important to at least cover the non-scalars. Although it results in better test coverage to have checks on the values before setting them, you're then relying on those checks to always work as intended.

**Comment by thetonymaster** [comment]
@bill-rich 

If I remember correctly, for non-scalar values we have them covered with the error checking. For scalar values, the package `utils` already does a check for this, the behavior of each of the functions prevents an error to occur if the value that comes from the struct is nil, it returns the default value, such as `0`, `""``false`, etc. I don't know if another behaviour would be desirable or this is ok for you, let us know.

**Comment by bill-rich** [comment]
@thetonymaster Sorry about that. I thought I'd seen some non-scalars that were missed, but I seem to have been mistaken on that. If you've got the non-scalars all checked and the are confident in properly handling scalars that are returned as the default, I don't think this is a blocking issue. More of an observation/suggestion to help make sure you don't run into problems in the future.

**Comment by bill-rich** [comment]
Looking through one more time, I did notice one other item that could be improved some. Don't need anything too crazy here, just want to exercise the primary functions and make sure they're behaving as expected. Sorry I missed this on the first pass.
- [ ] Add additional acceptance tests for updating, importing and checking that the expected values are returned.

**Comment by thetonymaster** [comment]
@bill-rich can you check it again, we added the update tests

**Comment by JonKohler** [comment]
hey @bill-rich - following up on Tony's comments, Tony and team have added the tests. How are we looking from your perspective?

**Comment by bill-rich** [comment]
Sorry @thetonymaster, I missed the import steps added in to the other tests. That solution looks good to me. 

On the `nutanix_network_security_rule` tests, it is important to be able to test the functions for all the resources included in the provider. Is `update` not currently working, or is it just a matter of how to implement the tests?

**Comment by JonKohler** [comment]
@bill-rich yes, we're moving the network_security_rule to a future release until we can get the test environment working for that feature. In the mean time, the other guys should be ready to go.

Was that the only remaining issue? or is there anything else still outstanding @bill-rich ?

**Comment by bill-rich** [comment]
That was the last question I had. I think we're set to go. Thanks for working through the changes!

**Comment by JonKohler** [comment]
ok fantastic @bill-rich - whats the next step? I haven't been through this process before. cc @cgriggs01

**Comment by cgriggs01** [comment]
This is great, we can move forward to the release process. 

@JonKohler I will reach out via email to explain the release phase. 

@bill-rich - you can close this issue now.

**Comment by JonKohler** [comment]
Thanks @cgriggs01 ! I'll just close this out since im already on the page, and I'll look for your email! cheers, jon
